gen/generate.go

```diff
@@ -115,94 +120,148 @@ func generateGenesisFiles(outDir string, protoVersion protocol.ConsensusVersion,
+				errorsChannel <- partkeyErr
+				return
 			}
-		} else {
-			// At this point either rootKeys is valid or rootkeyErr != nil
-			// Likewise, either partkey is valid or partkeyErr != nil
-			if rootkeyErr != nil {
-				os.Remove(wfilename)
-
-				rootDB, err = db.MakeErasableAccessor(wfilename)
-				if err != nil {
-					err = fmt.Errorf("couldn't open root DB accessor %s: %v", wfilename, err)
-				} else {
-					root, err = account.GenerateRoot(rootDB)
+
+			if rootkeyErr == nil && partkeyErr == nil {
+				if verbose {
+					verbosedOutput <- fmt.Sprintln("Reusing existing wallet:", wfilename, pfilename)
 				}
-				if err != nil {
+			} else {
+				// At this point either rootKeys is valid or rootkeyErr != nil
+				// Likewise, either partkey is valid or partkeyErr != nil
+				if rootkeyErr != nil {
 					os.Remove(wfilename)
-					return
+
+					rootDB, err = db.MakeErasableAccessor(wfilename)
+					if err != nil {
+						err = fmt.Errorf("couldn't open root DB accessor %s: %v", wfilename, err)
+					} else {
+						root, err = account.GenerateRoot(rootDB)
+					}
+					if err != nil {
+						os.Remove(wfilename)
+						errorsChannel <- err
+						return
+					}
+					if verbose {
+						verbosedOutput <- fmt.Sprintf("Created new rootkey: %s", wfilename)
+					}
+					atomic.AddInt64(&rootKeyCreated, 1)
 				}
-				if verbose {
-					fmt.Printf("Created new rootkey: %s\n", wfilename)
+
+				if partkeyErr != nil && wallet.Online == basics.Online {
+					os.Remove(pfilename)
+
+					partDB, err = db.MakeErasableAccessor(pfilename)
+					if err != nil {
+						err = fmt.Errorf("couldn't open participation DB accessor %s: %v", pfilename, err)
+						os.Remove(pfilename)
+						errorsChannel <- err
+						return
+					}
+
+					part, err = account.FillDBWithParticipationKeys(partDB, root.Address(), basics.Round(firstWalletValid), basics.Round(lastWalletValid), partKeyDilution)
+					if err != nil {
+						err = fmt.Errorf("could not generate new participation file %s: %v", pfilename, err)
+						os.Remove(pfilename)
+						errorsChannel <- err
+						return
+					}
+					if verbose {
+						verbosedOutput <- fmt.Sprintf("Created new partkey: %s", pfilename)
+					}
+					atomic.AddInt64(&partKeyCreated, 1)
 				}
-				rootKeyCreated++
 			}
 
-			if partkeyErr != nil && wallet.Online == basics.Online {
-				os.Remove(pfilename)
+			var data basics.AccountData
+			data.Status = wallet.Online
+			data.MicroAlgos.Raw = wallet.Stake
+			if wallet.Online == basics.Online {
+				data.VoteID = part.VotingSecrets().OneTimeSignatureVerifier
+				data.SelectionID = part.VRFSecrets().PK
+				data.VoteFirstValid = part.FirstValid
+				data.VoteLastValid = part.LastValid
+				data.VoteKeyDilution = part.KeyDilution
+			}
 
-				partDB, err = db.MakeErasableAccessor(pfilename)
-				if err != nil {
-					err = fmt.Errorf("couldn't open participation DB accessor %s: %v", pfilename, err)
-					os.Remove(pfilename)
-					return
-				}
+			writeMu.Lock()
```

Comment:
side thoughts: interesting if it is possible in go pre-allocate map and assign ranges to every goroutine so that it can write to maps without locking

Reply:
not with a map; but it's doable with a preallocated array.